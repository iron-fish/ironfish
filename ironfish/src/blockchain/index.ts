/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import Blockchain from './blockchain'
import { NullifierHasher, Nullifier } from './nullifiers'

export { NullifierHasher, Nullifier }
export * from './blockchain'
export default Blockchain
