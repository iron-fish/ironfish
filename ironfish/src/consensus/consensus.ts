/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { TransactionVersion } from '../primitives/transaction'

export type ActivationSequence = number | 'never'

export type ConsensusParameters = {
  /**
   * When adding a block, the block can be this amount of seconds into the future
   * without rejecting it
   */
  allowedBlockFutureSeconds: number

  /**
   * The amount of coins in the genesis block
   */
  genesisSupplyInIron: number

  /**
   * The average time that all blocks should be mined
   */
  targetBlockTimeInSeconds: number

  /**
   * The time range when difficulty and target not change
   */
  targetBucketTimeInSeconds: number

  /**
   * Max block size
   */
  maxBlockSizeBytes: number

  /**
   * The minimum fee that a transaction must have to be accepted
   */
  minFee: number

  /**
   * The block height that enables the use of V2 transactions instead of V1
   */
  enableAssetOwnership: ActivationSequence

  /**
   * Before upgrade we have block timestamp smaller than previous block. After this
   * block we enforce the block timestamps in the sequential order as the block sequences.
   */
  enforceSequentialBlockTime: ActivationSequence
}

export class Consensus {
  readonly parameters: ConsensusParameters

  constructor(parameters: ConsensusParameters) {
    this.parameters = parameters
  }

  isActive(upgrade: ActivationSequence, sequence: number): boolean {
    if (upgrade === 'never') {
      return false
    }
    return Math.max(1, sequence) >= upgrade
  }

  getActiveTransactionVersion(sequence: number): TransactionVersion {
    if (this.isActive(this.parameters.enableAssetOwnership, sequence)) {
      return TransactionVersion.V2
    } else {
      return TransactionVersion.V1
    }
  }
}

export class TestnetConsensus extends Consensus {
  constructor(parameters: ConsensusParameters) {
    super(parameters)
  }
}
