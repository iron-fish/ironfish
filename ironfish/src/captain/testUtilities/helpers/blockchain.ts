/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { SerializedTestTransaction, TestStrategy, TestTransaction } from '../strategy'
import Blockchain from '../../../blockchain'
import Block from '../../../blockchain/block'
import BlockHeader, { BlockHash } from '../../../blockchain/blockheader'
import { Target } from '../../../blockchain/target'
import { makeDb, makeDbName } from './storage'
import { RangeHasher } from '../../../merkletree/hasher'
import { createRootLogger } from '../../../logger'

/**
 * Make a block with a hash consisting of the given digit,
 * the previous hash consisting of the next digit, and the start and
 * end numbers of a sequence of notes in the block.
 *
 * Note: The resulting block is suitable for use on a blockchain.BlockChain,
 * but will fail if you try adding it to an anchorchain without some extra
 * massaging of the values.
 *
 * Specifically, the nullifier commitment does not have a correct value against
 * the value in the tree. The note commitment should match up, though it depends
 * exactly how the tree was initially produced.
 *
 * Most notably, a block created with this function will not go onto a chain
 * created with makeChain or makeCaptain. You are probably better off using
 * makeNextBlock from the anchorChain test utilities instead.
 */
export function makeFakeBlock(
  strategy: TestStrategy,
  previousHash: BlockHash,
  hash: BlockHash,
  sequence: number,
  start: number,
  end: number,
  timestamp?: Date,
): Block<string, string, TestTransaction, string, string, SerializedTestTransaction> {
  const transactions = []
  for (let i = start; i < end; i++) {
    transactions.push(new TestTransaction(true, [String(i)], 1))
  }

  const minersReward = strategy.miningReward(BigInt(sequence))
  const transactionFee = -1 * (end - start + minersReward)
  const transactionFeeTransaction = new TestTransaction(true, [String(end)], transactionFee)
  transactions.push(transactionFeeTransaction)

  const graffiti = Buffer.alloc(32)
  graffiti.write('fake block')

  const header = new BlockHeader(
    strategy,
    BigInt(sequence),
    previousHash,
    {
      commitment: `1-${end}`,
      size: end,
    },
    { commitment: Buffer.alloc(32), size: 1 },
    fakeMaxTarget(),
    0,
    timestamp ? timestamp : new Date(1598970000000 + hash[0]),
    BigInt(transactionFee),
    graffiti,
  )

  return new Block(header, transactions)
}

/**
 * Make a block hash with the hash set to the given digit
 */
export function fakeMaxTarget(): Target {
  return new Target(BigInt(2) ** BigInt(256) - BigInt(1))
}

/**
 * Make a block hash with the hash set to the given digit
 */
export function blockHash(digit: number): BlockHash {
  const hash = Buffer.alloc(32)
  hash[0] = digit
  return hash
}

/**
 * Make a nullifier with the hash set to the given digit.
 */
export function makeNullifier(digit: number): BlockHash {
  const hash = Buffer.alloc(32)
  hash[0] = digit
  return hash
}

export async function makeBlockchain(): Promise<
  Blockchain<
    string,
    string,
    TestTransaction<string>,
    string,
    string,
    SerializedTestTransaction<string>
  >
> {
  const name = makeDbName()
  const database = makeDb(name)

  const strategy = new TestStrategy(new RangeHasher())
  const chain = await Blockchain.new(database, strategy, createRootLogger())

  await database.open()
  return chain
}
