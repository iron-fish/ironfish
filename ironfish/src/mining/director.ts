/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import LeastRecentlyUsed from 'blru'
import { BufferSet } from 'buffer-map'
import { Account } from '../account'
import { Assert } from '../assert'
import { Blockchain } from '../blockchain'
import { Event } from '../event'
import { createRootLogger, Logger } from '../logger'
import { MemPool } from '../memPool'
import { Block } from '../primitives/block'
import { BlockHash, BlockHeader, BlockHeaderSerde } from '../primitives/blockheader'
import { Target } from '../primitives/target'
import { Transaction } from '../primitives/transaction'
import { Strategy } from '../strategy'
import { submitMetric } from '../telemetry'
import { ErrorUtils, GraffitiUtils } from '../utils'

/**
 * Number of transactions we are willing to store in a single block.
 */
const MAX_TRANSACTIONS_PER_BLOCK = 10
const MINING_DIFFICULTY_CHANGE_TIMEOUT = 10000

type DirectorState = { type: 'STARTED' } | { type: 'STOPPED' }

/**
 * Responsible for directing miners about which block to mine.
 *
 * Listens for changes to the blockchain head and emits a 'onBlockToMine' event
 * for each one.
 *
 * @typeParam E Note element stored in transactions and the notes Merkle Tree
 * @typeParam H the hash of an `E`. Used for the internal nodes and root hash
 *              of the notes Merkle Tree
 * @typeParam T Type of a transaction stored on the Blockchain
 * @typeParam ST The serialized format of a `T`. Conversion between the two happens
 *               via the `strategy`.
 */
export class MiningDirector {
  readonly chain: Blockchain
  readonly memPool: MemPool

  /**
   * The event creates a block header with loose transactions that have been
   * submitted to the network. It then waits for one of the miners to send it
   * a randomness value for that block. If one arrives, the block is reconstructed,
   * gossiped, and added to the local tree.
   */
  onBlockToMine = new Event<
    [{ miningRequestId: number; bytes: Buffer; target: Target; sequence: number }]
  >()

  /**
   * Emitted when a new block has been mined
   */
  onNewBlock = new Event<[Block]>()

  /**
   * The chain strategy used to calculate miner's fees.
   */
  strategy: Strategy

  /**
   * Serde to convert block headers to jsonable objects.
   */
  blockHeaderSerde: BlockHeaderSerde

  /**
   * Reference blocks that we most recently emitted for miners to mine.
   */
  recentBlocks: LeastRecentlyUsed<number, Block>

  /**
   * Block currently being generated by the director. Nulled out after
   * the miner's fee is generated. (It will be set to null while
   * retrying)
   */
  currentBlockUnderConstruction: BlockHash | null = null

  /**
   * Next block to construct after currentBlockUnderConstruction finishes.
   */
  nextBlockToConstruct: BlockHash | null = null

  /**
   * The value to set on the `graffiti` field of newly generated blocks.
   */
  private _blockGraffiti: string
  get blockGraffiti(): string {
    return this._blockGraffiti
  }

  /**
   * The private spending key for this miner. This is used to construct the
   * miner's fee transaction for the block.
   */
  private _minerAccount: Account | null
  get minerAccount(): Account | null {
    return this._minerAccount
  }

  /**
   * Logger instance used in place of console logs
   */
  logger: Logger

  /**
   * Setting an interval every 10 seconds to re-calculate the target for the
   * currentBlock based on updated timestamp
   */
  miningDifficultyChangeTimeout: null | ReturnType<typeof setTimeout>

  private _state: Readonly<DirectorState> = { type: 'STOPPED' }

  get state(): Readonly<DirectorState> {
    return this._state
  }

  setState(state: Readonly<DirectorState>): void {
    this._state = state
  }

  /**
   * Identifier for each request of blocks that gets sent to miners. This
   * increases monotonically and allows director to figure out which
   * block it is receiving randomness for.
   */
  private miningRequestId: number

  /**
   * How many blocks the miner has found
   */
  blocksMined = 0

  /**
   * How many blocks the miner has found
   */
  miners = 0

  /**
   * Should the miner mine, even if the chain is not synced
   */
  force: boolean

  constructor(options: {
    chain: Blockchain
    memPool: MemPool
    strategy: Strategy
    logger?: Logger
    graffiti?: string
    account?: Account
    force?: boolean
  }) {
    const logger = options.logger || createRootLogger()

    this.chain = options.chain
    this.memPool = options.memPool
    this.strategy = options.strategy
    this.logger = logger.withTag('director')

    this.force = options.force || false
    this._blockGraffiti = ''
    this._minerAccount = null
    this.blockHeaderSerde = new BlockHeaderSerde(options.strategy)
    this.miningDifficultyChangeTimeout = null
    this.miningRequestId = 0
    this.recentBlocks = new LeastRecentlyUsed(50)

    if (options.graffiti) {
      this.setBlockGraffiti(options.graffiti)
    }

    if (options.account) {
      this.setMinerAccount(options.account)
    }

    this.chain.onConnectBlock.on((block) => {
      // setTimeout for deadlock from newBlock()
      setTimeout(() => {
        void this.onChainHeadChange(block.header)
      })
    })
    this.chain.onSynced.on(() => this.onChainSynced())
  }

  async start(): Promise<void> {
    this.setState({ type: 'STARTED' })
    this.logger.debug('Mining director is running')

    if (!this.chain.synced && !this.force) {
      this.logger.debug(`Aborting mining because we are still syncing`)
      return
    }

    if (!this.chain.head) {
      this.logger.debug(`Aborting mining because we have no genesis block`)
      return
    }

    await this.generateBlockToMine(this.chain.head.hash)
  }

  isStarted(): boolean {
    return this.state.type === 'STARTED'
  }

  setBlockGraffiti(graffiti: string): void {
    this._blockGraffiti = graffiti
  }

  setMinerAccount(account: Account | null): void {
    this._minerAccount = account
  }

  /**
   * Event listener hooked up to changes in Blockchain.
   *
   * When a new head is received it:
   *  * adds any transactions that we were attempting to mine back to the pool
   *  * Creates a new block with transactions from the pool
   *  * emits the header of the new block to any listening miners
   *  * stores block until either the head changes again or it is succesfully mined
   *
   * @param newChainHead The hash of the new head of the chain
   * @event onBlockToMine header of a new block that needs to have its randomness mined
   */
  async onChainHeadChange(head: BlockHeader): Promise<void> {
    if (!this.isStarted()) {
      return
    }

    if (!this.chain.synced && !this.force) {
      return
    }

    try {
      await this.generateBlockToMine(head.hash)
    } catch (err: unknown) {
      this.logger.error(err)
    }
  }

  async onChainSynced(): Promise<void> {
    if (!this.isStarted()) {
      return
    }

    Assert.isNotNull(this.chain.head)
    await this.generateBlockToMine(this.chain.head.hash)
  }

  async generateBlockToMine(chainHead: BlockHash): Promise<void> {
    // Are we already mining that block
    if (
      this.currentBlockUnderConstruction !== null &&
      chainHead.equals(this.currentBlockUnderConstruction)
    ) {
      return
    }

    // If we're already generating a block, update the next block to generate and exit
    this.nextBlockToConstruct = chainHead
    if (this.currentBlockUnderConstruction !== null) {
      return
    }

    // Continue generating while we have a new block to generate
    while (this.nextBlockToConstruct !== null && this.isStarted()) {
      this.currentBlockUnderConstruction = this.nextBlockToConstruct
      this.nextBlockToConstruct = null

      if (this.miningDifficultyChangeTimeout) {
        clearTimeout(this.miningDifficultyChangeTimeout)
      }

      let blockData = null

      try {
        blockData = await this.constructTransactionsAndFees(this.currentBlockUnderConstruction)
      } catch (error: unknown) {
        this.logger.debug(
          `An error occurred while creating the new block ${ErrorUtils.renderError(error)}`,
        )
      }

      if (blockData === null) {
        continue
      }

      const [minersFee, blockTransactions] = blockData
      await this.constructAndMineBlockWithRetry(minersFee, blockTransactions)
    }

    // No longer generating a block
    this.currentBlockUnderConstruction = null
  }

  async constructAndMineBlockWithRetry(
    minersFee: Transaction,
    blockTransactions: Transaction[],
  ): Promise<void> {
    if (!this.isStarted()) {
      return
    }

    const canRetry = await this.constructAndMineBlock(minersFee, blockTransactions)
    // The current mining target is already at the initial - no need to try to lower it
    if (!canRetry) {
      return
    }

    if (this.miningDifficultyChangeTimeout) {
      clearTimeout(this.miningDifficultyChangeTimeout)
    }
    this.miningDifficultyChangeTimeout = setTimeout(() => {
      void this.constructAndMineBlockWithRetry(minersFee, blockTransactions)
    }, MINING_DIFFICULTY_CHANGE_TIMEOUT)
  }

  /**
   * Construct the transactions, header and miner fees used by constructAndMineBlock
   *
   * @param newChainHead The hash of the new head of the chain
   */
  async constructTransactionsAndFees(
    newChainHead: BlockHash,
  ): Promise<[Transaction, Transaction[]]> {
    if (!this._minerAccount) {
      throw Error('No miner account found to construct the transaction')
    }

    const blockTransactions = []

    // Fetch all transactions for the block
    for await (const transaction of this.getTransactions()) {
      if (blockTransactions.length >= MAX_TRANSACTIONS_PER_BLOCK) {
        break
      }

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      blockTransactions.push(transaction)
    }

    // Sum the transaction fees
    let totalTransactionFees = BigInt(0)
    const transactionFees = await Promise.all(blockTransactions.map((t) => t.fee()))
    for (const transactionFee of transactionFees) {
      totalTransactionFees += transactionFee
    }

    const blockHeader = await this.chain.getHeader(newChainHead)
    if (!blockHeader) {
      // Chain normally has a header for a heaviestHead. Block could be removed
      // if a predecessor is proven invalid while this task is running. (unlikely but possible)
      throw Error('No header for the new block')
    }

    const minersFee = await this.strategy.createMinersFee(
      totalTransactionFees,
      blockHeader.sequence + 1,
      this._minerAccount.spendingKey,
    )

    return [minersFee, blockTransactions]
  }

  /**
   * Construct a new block and send it out to miners.
   *
   * This is called both when the chain head changes and
   * when the timeout for mining a block at its current difficulty
   * expires.
   *
   * @param newChainHead The hash of the new head of the chain
   * @returns a promise that resolves to a boolean. Boolean returns
   * true if mining that block can be retried with a lower difficulty
   */
  async constructAndMineBlock(
    minersFee: Transaction,
    blockTransactions: Transaction[],
  ): Promise<boolean> {
    const newBlock = await this.chain.newBlock(
      blockTransactions,
      minersFee,
      GraffitiUtils.fromString(this.blockGraffiti),
    )

    this.logger.debug(
      `Current block  ${newBlock.header.sequence}, has ${newBlock.transactions.length} transactions`,
    )

    // For mining, we want a serialized form of the header without the randomness on it
    const target = newBlock.header.target
    this.logger.debug('target set to', target.asBigInt())
    const asBuffer = newBlock.header.serializePartial()
    this.miningRequestId++

    this.logger.debug(
      `Emitting a new block ${newBlock.header.sequence} to mine as request ${this.miningRequestId}`,
    )

    await this.onBlockToMine.emitAsync({
      bytes: asBuffer,
      target,
      miningRequestId: this.miningRequestId,
      sequence: newBlock.header.sequence,
    })

    this.recentBlocks.set(this.miningRequestId, newBlock)

    const canRetry = target.asBigInt() < Target.maxTarget().asBigInt()
    return canRetry
  }

  /**
   * Called when a block has been successfully mined.
   *
   * To reduce cost of communication with miners, only the randomness for
   * the new block is passed in. It is set on the block we have stored locally
   * and verified.
   *
   * The new block is added to the chain and sent out to be gossip'd.
   *
   * @param randomness The randomness to be set for the new block
   */
  async successfullyMined(randomness: number, miningRequestId: number): Promise<MINED_RESULT> {
    const block = this.recentBlocks.get(miningRequestId)
    if (!block) {
      return MINED_RESULT.UNKNOWN_REQUEST
    }

    // TODO(jason): This is temporary to avoid the race condition here of
    // finding multiple results for the same block. It'll mutate the block that
    // is shared by all other calls in recentBlocks then rely on that value
    // while doing async work. Then other calls can come in at the same time. We
    // should remove this when we can put a lock around this whole function's
    // code
    this.recentBlocks.remove(miningRequestId)

    if (!this.chain.head || !block.header.previousBlockHash.equals(this.chain.head.hash)) {
      this.logger.debug('Discarding block that no longer attaches to heaviest head')
      return MINED_RESULT.CHAIN_CHANGED
    }

    block.header.randomness = randomness
    const validation = await this.chain.verifier.verifyBlock(block)

    if (!validation.valid) {
      this.logger.warn('Discarding invalid block', validation.reason)
      return MINED_RESULT.INVALID_BLOCK
    }

    this.logger.info(
      `Successfully mined block ${block.header.hash.toString('hex')} (${
        block.header.sequence
      }) has ${block.transactions.length} transactions`,
    )

    this.blocksMined++

    const { isAdded, reason } = await this.chain.addBlock(block)

    if (!isAdded) {
      this.logger.error(`Failed to add mined block to chain with reason ${String(reason)}`)
      return MINED_RESULT.ADD_FAILED
    }

    this.onNewBlock.emit(block)

    submitMetric({
      name: 'minedBlock',
      fields: [
        {
          name: 'difficulty',
          type: 'integer',
          value: Number(block.header.target.toDifficulty()),
        },
        {
          name: 'sequence',
          type: 'integer',
          value: Number(block.header.sequence),
        },
      ],
    })

    return MINED_RESULT.SUCCESS
  }

  /**
   * clears the timeout to queue up more mining jobs with recalculated target
   */
  shutdown(): void {
    this.setState({ type: 'STOPPED' })

    if (this.miningDifficultyChangeTimeout) {
      clearTimeout(this.miningDifficultyChangeTimeout)
    }
  }

  protected async *getTransactions(): AsyncGenerator<Transaction> {
    const nullifiers = new BufferSet()

    for (const transaction of this.memPool.get()) {
      let conflicted = false

      for (const spend of transaction.spends()) {
        if (nullifiers.has(spend.nullifier)) {
          conflicted = true
          break
        }

        nullifiers.add(spend.nullifier)

        if (await this.chain.nullifiers.contains(spend.nullifier)) {
          conflicted = true
          break
        }
      }

      if (conflicted) {
        continue
      }

      yield transaction
    }
  }
}

export enum MINED_RESULT {
  UNKNOWN_REQUEST = 'UNKNOWN_REQUEST',
  CHAIN_CHANGED = 'CHAIN_CHANGED',
  INVALID_BLOCK = 'INVALID_BLOCK',
  ADD_FAILED = 'ADD_FAILED',
  SUCCESS = 'SUCCESS',
}
