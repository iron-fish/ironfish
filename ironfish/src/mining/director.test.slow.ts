/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { VerificationResultReason } from '../consensus'
import { waitForEmit } from '../event'
import { Target } from '../primitives/target'
import PartialBlockHeaderSerde from '../serde/PartialHeaderSerde'
import {
  createNodeTest,
  useAccountFixture,
  useMinerBlockFixture,
  useMinersTxFixture,
  useTxFixture,
} from '../testUtilities'
import { makeBlockAfter } from '../testUtilities/helpers/blockchain'
import { flushTimeout } from '../testUtilities/helpers/tests'
import { isTransactionMine } from '../testUtilities/helpers/transaction'
import { MINED_RESULT } from './director'

describe('Mining director', () => {
  const nodeTest = createNodeTest()

  beforeEach(async () => {
    await nodeTest.node.miningDirector.start()
  })

  afterEach(() => {
    nodeTest.node.miningDirector.shutdown()
  })

  describe('Before mining', () => {
    it('creates a new block to be mined when chain head changes', async () => {
      const { chain, miningDirector } = nodeTest.node
      miningDirector.force = true

      const generateBlockSpy = jest
        .spyOn(miningDirector, 'generateBlockToMine')
        .mockReturnValue(Promise.resolve())

      // onSynced can trigger while generating fixtures or if this test is run in isolation,
      // which would call generateBlockToMine twice, so we can clear the listener to ensure it
      // will only be called once.
      chain.onSynced.clear()

      const previous = await useMinerBlockFixture(chain, 2)

      await expect(chain).toAddBlock(previous)
      await flushTimeout()

      expect(generateBlockSpy).toBeCalledTimes(1)
      expect(generateBlockSpy.mock.calls[0][0].equals(previous.header.hash)).toBe(true)
    }, 10000)

    it('creates block and starts mining', async () => {
      const { chain, accounts, miningDirector } = nodeTest.node

      const account = await accounts.createAccount('')
      miningDirector.setMinerAccount(account)

      const spy = jest
        .spyOn(miningDirector, 'constructAndMineBlockWithRetry')
        .mockReturnValue(Promise.resolve())

      await miningDirector.generateBlockToMine(chain.head.hash)

      expect(spy).toBeCalledTimes(1)
      const [minersFee, transactions] = spy.mock.calls[0]
      expect(transactions).toHaveLength(0)
      expect(isTransactionMine(minersFee, account)).toBe(true)
    }, 10000)

    it('adds transactions from the mempool', async () => {
      const { node, miningDirector, chain } = nodeTest

      const account = await useAccountFixture(node.accounts, 'a')
      miningDirector.setMinerAccount(account)
      miningDirector.force = true

      const previous = await useMinerBlockFixture(chain, 2, account, node.accounts)
      await expect(chain).toAddBlock(previous)
      await node.accounts.updateHead()

      const transaction = await useTxFixture(node.accounts, account, account)

      expect(node.memPool.size()).toBe(0)
      await miningDirector.memPool.acceptTransaction(transaction)
      expect(node.memPool.size()).toBe(1)

      const spy = jest
        .spyOn(miningDirector, 'constructAndMineBlockWithRetry')
        .mockReturnValue(Promise.resolve())

      await miningDirector.generateBlockToMine(chain.head.hash)

      expect(spy).toBeCalledTimes(1)
      const transactions = spy.mock.calls[0][1]
      expect(transactions).toHaveLength(1)
      expect(isTransactionMine(transactions[0], account)).toBe(true)
      expect(node.memPool.size()).toBe(1)
    }, 25000)

    it('should emit block to be mined', async () => {
      // This test is testing the partially constructed
      // block that's created to be mined is reasonably
      // correct when emitted from MiningDirector.onBlockToMine
      const { chain, miningDirector, accounts } = nodeTest.node
      const account = await accounts.createAccount('')
      nodeTest.strategy.disableMiningReward()
      miningDirector.force = true
      miningDirector.setBlockGraffiti('testing')
      miningDirector.setMinerAccount(account)

      const now = Date.now()

      // Freeze time so we can predict the target
      jest.spyOn(global.Date, 'now').mockReturnValue(now)
      jest.spyOn(miningDirector.onBlockToMine, 'emit')

      const promise = waitForEmit(miningDirector.onBlockToMine)
      const previous = await makeBlockAfter(chain, chain.head)

      await expect(chain).toAddBlock(previous)
      const [event] = await promise

      const partial = new PartialBlockHeaderSerde().deserialize(event.bytes)

      expect(event.target.targetValue).toEqual(
        Target.calculateTarget(new Date(now), previous.header.timestamp, previous.header.target)
          .targetValue,
      )

      expect(partial.previousBlockHash.equals(previous.header.hash)).toBe(true)
      expect(partial.minersFee).toEqual(BigInt(0))
      expect(partial.timestamp.valueOf()).toEqual(now)
    }, 15000)
  })

  describe('During Mining', () => {
    beforeEach(() => {
      jest.useFakeTimers()
      jest.setTimeout(15000000)
    })

    afterEach(() => {
      jest.useRealTimers()
    })

    it('should recalculate block after interval', async () => {
      const { node, miningDirector } = nodeTest
      const account = await useAccountFixture(node.accounts, 'a')
      const tx = await useMinersTxFixture(nodeTest.node.accounts, account)

      const spy = jest.spyOn(miningDirector, 'constructAndMineBlock').mockResolvedValue(false)
      expect(spy).toBeCalledTimes(0)

      await miningDirector.constructAndMineBlockWithRetry(tx, [])
      expect(spy).toBeCalledTimes(1)

      // Should not retry
      jest.advanceTimersByTime(11000)
      expect(spy).toBeCalledTimes(1)

      spy.mockResolvedValue(true)
      await miningDirector.constructAndMineBlockWithRetry(tx, [])
      expect(spy).toBeCalledTimes(2)

      // should retry now
      jest.advanceTimersByTime(11000)
      expect(spy).toBeCalledTimes(3)
    })

    it('retries calculating target if target is not at max', async () => {
      const { node, miningDirector } = nodeTest
      const account = await useAccountFixture(node.accounts, 'a')
      const tx = await useMinersTxFixture(nodeTest.node.accounts, account)

      const targetSpy = jest.spyOn(Target, 'calculateTarget')

      targetSpy.mockReturnValueOnce(Target.maxTarget())
      await expect(miningDirector.constructAndMineBlock(tx, [])).resolves.toBe(false)

      targetSpy.mockReturnValueOnce(
        Target.fromDifficulty(Target.minDifficulty() + BigInt(10000000000)),
      )
      await expect(miningDirector.constructAndMineBlock(tx, [])).resolves.toBe(true)
    })
  })

  describe('After mining', () => {
    it('discards block if not in recentBlocks', async () => {
      const { miningDirector } = nodeTest

      await expect(miningDirector.successfullyMined(0, 0)).resolves.toBe(
        MINED_RESULT.UNKNOWN_REQUEST,
      )
    })

    it('discards block if chain changed', async () => {
      const { strategy, chain, node, miningDirector } = nodeTest
      strategy.disableMiningReward()

      const blockA1 = await makeBlockAfter(chain, chain.genesis)
      const blockA2 = await makeBlockAfter(chain, chain.genesis)

      node.miningDirector.recentBlocks.set(2, blockA1)
      node.miningDirector.recentBlocks.set(3, blockA2)

      await expect(miningDirector.successfullyMined(1, 2)).resolves.toBe(MINED_RESULT.SUCCESS)

      await expect(miningDirector.successfullyMined(1, 3)).resolves.toBe(
        MINED_RESULT.CHAIN_CHANGED,
      )
    })

    it('discards block if not valid', async () => {
      const { strategy, chain, node, miningDirector } = nodeTest
      strategy.disableMiningReward()

      const block = await makeBlockAfter(chain, chain.genesis)
      node.miningDirector.recentBlocks.set(2, block)

      jest
        .spyOn(chain.verifier, 'verifyBlock')
        .mockResolvedValue({ valid: false, reason: VerificationResultReason.INVALID_TARGET })

      await expect(miningDirector.successfullyMined(1, 2)).resolves.toBe(
        MINED_RESULT.INVALID_BLOCK,
      )
    })

    it('discard block if cannot add to chain', async () => {
      const { strategy, chain, node, miningDirector } = nodeTest
      strategy.disableMiningReward()

      const block = await makeBlockAfter(chain, chain.genesis)
      node.miningDirector.recentBlocks.set(2, block)

      jest.spyOn(chain, 'addBlock').mockResolvedValue({
        isAdded: false,
        reason: VerificationResultReason.INVALID_TARGET,
        score: 0,
      })

      await expect(miningDirector.successfullyMined(1, 2)).resolves.toBe(
        MINED_RESULT.ADD_FAILED,
      )
    })

    it('adds block on successful mining', async () => {
      const { strategy, chain, node } = nodeTest
      strategy.disableMiningReward()

      const onNewBlockSpy = jest.spyOn(node.miningDirector.onNewBlock, 'emit')

      const block = await useMinerBlockFixture(chain, 2)
      node.miningDirector.recentBlocks.set(1, block)

      await node.miningDirector.successfullyMined(5, 1)
      expect(onNewBlockSpy).toBeCalledWith(block)
    })
  })
})
