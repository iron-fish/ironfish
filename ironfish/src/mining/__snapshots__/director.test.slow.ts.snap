// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mining director adds transactions from the queue to a new block to be mined 1`] = `
Object {
  "graffiti": "0000000000000000000000000000000000000000000000000000000000000000",
  "minersFee": "-100",
  "noteCommitment": Object {
    "commitment": "0-miners note 100",
    "size": 45,
  },
  "nullifierCommitment": Object {
    "commitment": "B08D0434ABFFFB68F6374100A5E11553519AB0270FFE18137669AFA18E22B519",
    "size": 18,
  },
  "previousBlockHash": "0900000000000000000000000000000000000000000000000000000000000000",
  "sequence": "10",
  "target": "115792089237316195423570985008687907853269984665640564039457584007913129639935",
  "timestamp": Any<Number>,
}
`;

exports[`Mining director creates a new block to be mined when chain head changes 2`] = `
Object {
  "graffiti": "0000000000000000000000000000000000000000000000000000000000000000",
  "minersFee": "-10",
  "noteCommitment": Object {
    "commitment": "0-miners note 10",
    "size": 41,
  },
  "nullifierCommitment": Object {
    "commitment": "1BA8B29F458F6FEBB8EB43097EEFCE7823CB306D56EDF4035B1849E3FECC0687",
    "size": 16,
  },
  "previousBlockHash": "0900000000000000000000000000000000000000000000000000000000000000",
  "sequence": "10",
  "target": "115792089237316195423570985008687907853269984665640564039457584007913129639935",
  "timestamp": Any<Number>,
}
`;

exports[`Mining director does not add invalid transactions to the block 1`] = `
Object {
  "graffiti": "0000000000000000000000000000000000000000000000000000000000000000",
  "minersFee": "-10",
  "noteCommitment": Object {
    "commitment": "0-miners note 10",
    "size": 41,
  },
  "nullifierCommitment": Object {
    "commitment": "1BA8B29F458F6FEBB8EB43097EEFCE7823CB306D56EDF4035B1849E3FECC0687",
    "size": 16,
  },
  "previousBlockHash": "0900000000000000000000000000000000000000000000000000000000000000",
  "sequence": "10",
  "target": "115792089237316195423570985008687907853269984665640564039457584007913129639935",
  "timestamp": Any<Number>,
}
`;
